plugins/source/azuredevops/codegen/recipes/base.go

```diff
@@ -0,0 +1,114 @@
+package recipes
+
+import (
+	"bytes"
+	"embed"
+	"fmt"
+	"go/format"
+	"os"
+	"path"
+	"reflect"
+	"runtime"
+	"text/template"
+
+	"github.com/cloudquery/plugin-sdk/codegen"
+	"github.com/cloudquery/plugin-sdk/schema"
+	"github.com/google/uuid"
+	"github.com/iancoleman/strcase"
+)
+
+//go:embed templates/*.go.tpl
+var templatesFS embed.FS
+
+type Resource struct {
+	Service    string
+	SubService string
+	Struct     interface{}
+	Table      *codegen.TableDefinition
+}
+
+func renderTemplate(name string, filePath string, data interface{}) error {
+	_, filename, _, ok := runtime.Caller(0)
+	if !ok {
+		return fmt.Errorf("failed to get caller information")
+	}
+	dir := path.Dir(filename)
+
+	tpl, err := template.New(name).Funcs(template.FuncMap{"ToCamel": strcase.ToCamel}).ParseFS(templatesFS, "templates/*.go.tpl")
+	if err != nil {
+		return fmt.Errorf("failed to parse azureDevops templates: %w", err)
+	}
+	tpl, err = tpl.ParseFS(codegen.TemplatesFS, "templates/*.go.tpl")
+	if err != nil {
+		return fmt.Errorf("failed to parse sdk template: %w", err)
+	}
+
+	var buff bytes.Buffer
+	if err := tpl.Execute(&buff, data); err != nil {
+		return fmt.Errorf("failed to execute template: %w", err)
+	}
+	outputPath := path.Join(dir, "../..", filePath)
+	outputDir := path.Dir(outputPath)
+	if err := os.MkdirAll(outputDir, os.ModePerm); err != nil {
+		return fmt.Errorf("failed to create directory %s: %w", outputDir, err)
+	}
+
+	content := buff.Bytes()
```

Comment:
extract write: https://github.com/cloudquery/cloudquery/blob/main/plugins/source/tailscale/codegen/resources/write.go

Reply:
Happy to extract it to make the code more readable